signac/core/jsoncollection.py

```diff
@@ -6,59 +6,97 @@
 This implements the JSON-backend for SyncedCollection API by
 implementing sync and load methods.
 """
-
 import os
 import json
 import errno
 import uuid
 
 from .synced_collection import SyncedCollection
+from .buffered_collection import BufferedSyncedCollection
 from .syncedattrdict import SyncedAttrDict
 from .synced_list import SyncedList
+from .buffered_collection import _get_filemetadata
+
+
+def get_namespace(class_name):
+    """Generate namespace for classname"""
+    return uuid.uuid5(uuid.NAMESPACE_URL, 'signac::'+class_name)
 
 
-class JSONCollection(SyncedCollection):
+class JSONCollection(BufferedSyncedCollection):
     """Implement sync and load using a JSON back end."""
 
     backend = __name__  # type: ignore
 
-    def __init__(self, filename=None, write_concern=False, **kwargs):
-        self._filename = os.path.realpath(filename) if filename is not None else None
-        self._write_concern = write_concern
+    def __init__(self, filename=None, data=None, write_concern=False, no_sync=False, **kwargs):
+        kwargs['data'] = data
         super().__init__(**kwargs)
         if (filename is None) == (self._parent is None):
             raise ValueError(
                 "Illegal argument combination, one of the two arguments, "
                 "parent or filename must be None, but not both.")
+        self.backend_kwargs['filename'] = None if filename is None else os.path.realpath(filename)
+        self.backend_kwargs['write_concern'] = write_concern
+        self.backend_kwargs['backend'] = self.backend
+        self._id = str(uuid.uuid5(get_namespace(type(self).__name__),
+                       self.backend_kwargs['filename'])) if filename else None
+        if not no_sync and data is not None:
+            self.sync()
 
     def _load(self):
         """Load the data from a JSON-file."""
         try:
-            with open(self._filename, 'rb') as file:
+            with open(self.backend_kwargs['filename'], 'rb') as file:
                 blob = file.read()
                 return json.loads(blob)
         except IOError as error:
             if error.errno == errno.ENOENT:
                 return None
 
-    def _sync(self):
+    def _sync(self, data=None):
         """Write the data to json file."""
```

Comment:
We have inconsistency with how JSON files are referenced. Here it's `json file` but above it was `JSON-file`. Let's be consistent. My preference for both cases is `JSON file`.